### PR TITLE
fix orbit controls with translated parent

### PR DESCRIPTION
**Pull request by marwie**
**Description**

Orbit controls don't work properly when parented to a non-root object in the hierarchy and orbiting behaves unexpectedly when parent object has some offset.

This PR replaces direct access to ``position`` and instead calls two new internal methods: ``getPosition`` and ``setPosition``. These in turn do check if the object is parented to the scene (if yes they just return the object's local position) and if not they convert from and to world position.

This contribution is funded by 🌵 [needle](https://needle.tools).

Example hierarchy:
20220218-145316_Fork
image

Behaviour before:
https://user-images.githubusercontent.com/5083203/154695840-4b344a61-c3c3-40de-bafb-ee7d237217ee.mp4

Behaviour with PR applied:
https://user-images.githubusercontent.com/5083203/154695884-5462c725-9ca2-47f8-a932-8c16db842efd.mp4

### CODE REVIEW

**Comment by Mugen87** [comment]
TBH, the common usage of `OrbitControls` is to use it with a camera that has no ancestor nodes. I understand your use case and also your fix however I only vote to merge this PR if more users bump into the same issue. In general, we want to keep the controls simple and focused on the most important features. That does of course mean certain edge cases are not honored.

**Comment by marwie** [comment]
Thanks for your reply. I understand your point. I'm wondering if you'd be more likely to vote if the PR would instead expose some sort of minimal API so I could apply make the fix for my usecase (e.g. overriding a setPosition and getPosition function)? Altough this would be less clear and transparent and maybe worse, would be interested in your thoughts on it or if you see a better option?

Alternatively the only way I see right now would be to fork OrbitControls and try to merge future fixes and improvements back into my modified version.

**Comment by hybridherbst** [comment]
I'm not sure if that counts ;) but I encountered this issue as well multiple times in the past.

Example cases:
- having an animation of a rollercoaster and wanting to orbit around the cart
- more generalized, having a smoothly animated target of sorts where you orbit around in and can also move a bit (so just pulling the orbit target along is not the right way, can lead to timing issues on fast motion etc.)

Both of the above would be trivial if OrbitControls would respect parent settings (after all, nothing states that "you can only use this on a root-level camera" and I'm not sure why that artifical limitation would be kept).

**Comment by LeviPesin** [comment]
I have also encountered the issue - it is very useful in many cases to place the camera directly in the GLTF model (if this model is a player model and the camera is the player camera). And this did not work with the OrbitControls.

File: examples/jsm/controls/OrbitControls.js
Comment (LeviPesin) [inline examples/jsm/controls/OrbitControls.js:379]: You can use `.copy()` here.

**Comment by hybridherbst** [comment]
As a simple test to show what this change fixes, and what usecases it allows for, I added this [here in misc_controls_orbit.html](https://github.com/mrdoob/three.js/blob/7ba77b3cc576a7862a106d573856742016309d79/examples/misc_controls_orbit.html#L94): 
```
if ( i == 0 ) {

    mesh.add( camera );
    camera.position.set( 1,1,1 );
    controls.target.copy( mesh.position );
    mesh.material = new THREE.MeshPhongMaterial( material );
    mesh.material.color.set( 'red' );

}
```
This is similar to the usecase @LeviPesin is describing above: the camera is parented to another scene object coming from an external file or just because it makes sense for the application.

Without this PR:
https://user-images.githubusercontent.com/2693840/161263348-d7053093-09c1-4168-856a-7c2b76b9a79e.mp4

With this PR:
https://user-images.githubusercontent.com/2693840/161263597-abcadcfd-1686-4c55-83ef-192db2f004c6.mp4

**Comment by WestLangley** [comment]
Instead, try this in your animation loop:

```js
controls.target.copy( object.position );
```
or
```js
controls.target.setFromMatrixPosition( object.matrixWorld );
```

**Comment by hybridherbst** [comment]
@WestLangley that's kind of missing the point of this PR – even if the object isn't moving (as in the example code I shared above) the result of orbiting a parented camera is currently incorrect and will be correct with this PR merged. Setting the target position every frame doesn't change that...

File: examples/jsm/controls/OrbitControls.js
Comment (Mugen87) [inline examples/jsm/controls/OrbitControls.js:372]: I wonder if we could simplify the PR and always use `getWorldPosition()` instead of introducing `getPosition()`. The additional overhead for the default case (no parent) is so small that there is no need for a different code path.

File: examples/jsm/controls/OrbitControls.js
Comment (Mugen87) [inline examples/jsm/controls/OrbitControls.js:169]: I think it's more clear to define something like `_position` in the closure scope above and reuse it in `update()`.

File: examples/jsm/controls/OrbitControls.js
Comment (Mugen87) [inline examples/jsm/controls/OrbitControls.js:381]: Instead of having `setPosition()`, can't we just always use `scope.object.worldToLocal()`?

File: examples/jsm/controls/OrbitControls.js
Comment (marwie) [reply examples/jsm/controls/OrbitControls.js:372]: I think that's a good point. I'll apply your suggestions later and test again

File: examples/jsm/controls/OrbitControls.js
Comment (LeviPesin) [inline examples/jsm/controls/OrbitControls.js:253]: Will this break if object has no parent?

Comment (marwie) [reply examples/jsm/controls/OrbitControls.js:253]: It always has a scene as a parent I think

Comment (LeviPesin) [reply examples/jsm/controls/OrbitControls.js:253]: I think it does not... Use `obj.parent ? obj.parent.worldToLocal( position ) : position.copy9 obj.position )`.

Comment (marwie) [reply examples/jsm/controls/OrbitControls.js:253]: When would the camera not be at least parented in a scene?

Comment (Mugen87) [reply examples/jsm/controls/OrbitControls.js:253]: Why don't you just use this?
```js
scope.object.worldToLocal( position );
```

Comment (Mugen87) [reply examples/jsm/controls/OrbitControls.js:253]: > When would the camera not be at least parented in a scene?

In almost all cases the camera is not added to the scene.

Comment (marwie) [reply examples/jsm/controls/OrbitControls.js:253]: > Why don't you just use this?
> 
> ```js
> scope.object.worldToLocal( position );
> ```

because then I get this behaviour:

https://user-images.githubusercontent.com/5083203/162046662-078dd9a1-2be5-4fd9-9cbb-3e5fa155469d.mp4

File: examples/jsm/controls/OrbitControls.js
Comment (Mugen87) [inline examples/jsm/controls/OrbitControls.js:253]: `scope.object.parent.worldToLocal( position )` does not update `scope.object.position`. Can you try this please:
```js
if ( scope.object.parent !== null ) {

    scope.object.worldToLocal( position );

}

scope.object.position.copy( position );
```

Comment (Mugen87) [inline examples/jsm/controls/OrbitControls.js:253]: `scope.object.parent.worldToLocal( position )` does not update `scope.object.position`. Can you try this please:
```js
if ( scope.object.parent !== null ) {

    scope.object.worldToLocal( position );

}

scope.object.position.copy( position );
```

Comment (Mugen87) [reply examples/jsm/controls/OrbitControls.js:253]: Or ideally just this:
```js
scope.object.worldToLocal( position );
scope.object.position.copy( position );
```

Comment (marwie) [reply examples/jsm/controls/OrbitControls.js:253]: It does both result in a flickering camera (as above)

``position`` is actually the object.position Vector3 no because of this few lines above:

```			const _position = object.position;

			return function update() {

				const position = scope.object.getWorldPosition( _position );
```

Comment (marwie) [reply examples/jsm/controls/OrbitControls.js:253]: Which means we can actually save this copy completely and just do:

```
				if(scope.object.parent)
					scope.object.parent.worldToLocal( position );
```

Comment (Mugen87) [reply examples/jsm/controls/OrbitControls.js:253]: > position is actually the object.position Vector3 i think because of this few lines above:

I'm afraid that's not true. `getWorldPosition()` returns a reference to the passed target vector.

Comment (Mugen87) [reply examples/jsm/controls/OrbitControls.js:253]: In any event, I see now why `scope.object.parent` has to be used when calling `worldToLocal()` 👍 .

Comment (marwie) [reply examples/jsm/controls/OrbitControls.js:253]: >> I'm afraid that's not true. getWorldPosition() returns a reference to the passed target vector.

Yes and the target vector is actually ``object.position``  because this is what we saved in ``const _position = object.position``. I probably misunderstood you [here](https://github.com/mrdoob/three.js/pull/23526#discussion_r843668853)

**Comment by Mugen87** [comment]
> because this is what we saved in const _position = object.position in the outer scope of update().

Oh, I've missed that line! I think it would be less confusing to not save the position reference in the closure scope. It's a bit confusing since we normally don't do that.

File: examples/jsm/controls/OrbitControls.js
Comment (Mugen87) [reply examples/jsm/controls/OrbitControls.js:253]: > because this is what we saved in const _position = object.position in the outer scope of update().

Oh, I've missed that line! I think it would be less confusing to not save the position reference in the closure scope. It's a bit confusing since we normally don't do that.

Comment (marwie) [reply examples/jsm/controls/OrbitControls.js:251]: Great, removed it and tested with ``camera without parent``, ``camera in scene root``, ``camera with parent offset``

**Review by Mugen87 — APPROVED** [review]

**Comment by looeee** [comment]
@marwie if you want a more full featured camera controls I recommend checking out @yomotsu's [camera-controls](https://github.com/yomotsu/camera-controls) package. I don't know if it works with translated parents but in general I have found it a good replacement when I bump into the limitations of `OrbitControls`. Which is pretty much always 😅

**Comment by marwie** [comment]
@looeee thanks for the recommendation, will take a look

**Comment by WestLangley** [comment]
>TBH, the common usage of OrbitControls is to use it with a camera that has no ancestor nodes. I understand your use case and also your fix however I only vote to merge this PR if more users bump into the same issue. In general, we want to keep the controls simple and focused on the most important features. That does of course mean certain edge cases are not honored.

Agreed.

Furthermore, what if the parent is rotated? This PR ignores part of the transform matrix.

I am not in favor of this PR. Use another controls, or hack your own copy. That is why the controls are part of the examples -- and not in core.

**Comment by marwie** [comment]
The PR is really simple in terms of code added or changed, doesn't really add much overhead as far as I can see and prevents certain usecases from breaking completely.
This is how a setup with rotated parent looks like btw:
image

**Comment by Mugen87** [comment]
@marwie Sorry, but @WestLangley is right. We want to focus on the most important use cases. Only with this policy we can keep the controls simple and manageable. Even if that means certain edge cases are not supported.

**Comment by LeviPesin** [comment]
This is not an edge case. It would be **very** useful to be able to use OrbitControls with translated parent.

**Comment by Mugen87** [comment]
Sorry but it is. I have seen a lot of user level code in the last years and never encountered this scenario.

**Comment by LeviPesin** [comment]
This scenario was not used only because it is not supported by the OrbitControls.

**Comment by Mugen87** [comment]
If so, we would have seen more error reports or feature requests in the past. Which did not happen.

**Comment by LeviPesin** [comment]
It was raised multiple times before adding the "camera cannot be a child of another object" to the documentation:
https://github.com/mrdoob/three.js/issues/15267
https://github.com/mrdoob/three.js/pull/15268
https://github.com/mrdoob/three.js/pull/16374
https://github.com/mrdoob/three.js/pull/16506
https://github.com/mrdoob/three.js/pull/19724

**Comment by Mugen87** [comment]
My opinion still stands. These issues were raised over many years and are not something frequent.

**Comment by LeviPesin** [comment]
They all (except the last one) were raised over half a year before adding that to the documentation.

**Comment by Mugen87** [comment]
Sorry, I don't it makes sense to continue this back and forth. I'm not going to change my mind here.

**Comment by CodyJasonBennett** [comment]
> I am not in favor of this PR. Use another controls, or hack your own copy. That is why the controls are part of the examples -- and not in core.

That's not how people use examples -- they import them alongside three. It's confusing to see the disparity in focus here and with loaders, for example. Are we misusing these modules? What's the policy in usage/modification to these?

> @marwie Sorry, but @WestLangley is right. We want to focus on the most important use cases. Only with this policy we can keep the controls simple and manageable. Even if that means certain edge cases are not supported.

Is this a matter of principle in supporting this use case or the complexity of this PR? I'm not sure if a follow-up would be appreciated or what else needs to be done if this were continued.

**Comment by LeviPesin** [comment]
About the complexity - this PR does not look very complex, but if it looks, even more simpler https://github.com/mrdoob/three.js/pull/16374 or https://github.com/mrdoob/three.js/pull/16506 can be used instead.

**Comment by hybridherbst** [comment]
This PR adds a grand total of 2 lines and changes a few in an example file that many, many people use and that is over 1200 lines long because it tries to (and succeeds, for the most part!) to be _the_ example for how to do OrbitControls. So much, that for almost every project it stays the right solution.
Here's the full diff for comparison: https://github.com/mrdoob/three.js/pull/23526/files.

There were changes based on what was requested, the question of "do other people need this" was answered positively with a good amount of users chiming in and referencing issues and PRs.

A clear usecase was given, loading a glTF file with a camera and attaching OrbitControls to that, which is currently broken. cc @donmccurdy 

A known existing bug (that was documented, turning it into a somehow "intentional limitation") is fixed by this PR.

What else is needed to see something as improvement to three.js? This is a serious question. I'd love to learn more about this - we were planning to contribute bug fixes and improvements back to three.js, but the experience with this PR here doesn't make that feel viable anymore. cc @mrdoob

I'm also now really confused about "core" vs. "examples" in regards to three; my understanding was that "core" needs to be polished, small, strict, and examples have a bit more leeway when it comes to implementation details. The history of this PR suggests otherwise though.

**TL;DR:** I find it really strange that a small bugfix like this can't be merged for confusing reasons.

**Comment by donmccurdy** [comment]
I understand how this looks like simply a small bugfix, but it's also an increase in the scope of OrbitControls — and the PR only partially implements that additional scope (to confirm: are rotated/scaled ancestors fully supported?), likely leaving three.js maintainers to deal with the bug reports and possibly implementation of the rest. If the rest of that scope is similarly trivial then (for myself at least) I wouldn't be opposed here.

> > I am not in favor of this PR. Use another controls, or hack your own copy. That is why the controls are part of the examples -- and not in core.
> 
> That's not how people use examples -- they import them alongside three. It's confusing to see the disparity in focus here and with loaders, for example. Are we misusing these modules? What's the policy in usage/modification to these?

It is hard to answer "what are the examples" in any universally coherent way. Some are basic, supporting common use cases and intentionally simple enough for end-user customization (e.g. most controls and helpers). Some are complex, aiming to be _the_ solution for a problem, and more challenging to customize (e.g. loaders). Other modules are drifting somewhere in the middle, generally increasing in complexity and maintenance cost over time.

There are good reasons why projects like [three-gpu-pathtracer](https://github.com/gkjohnson/three-gpu-pathtracer/) or [three-camera-controls](https://github.com/yomotsu/camera-controls) or [three-pathfinding](https://github.com/donmccurdy/three-pathfinding) are not part of this repository. At a certain level of complexity it would be too much burden for core maintainers, or — if there is a dedicated maintainer — they decide that maintenance in a dedicated repository is simplify easier, given full freedom to control the build process, unit tests, type-checking, dependencies, and so on.

I think the conflict here is the tension of trying to keep a popular example like OrbitControls from drifting slowly into the latter category. Sorry for the confusion in the grey area. 😔

**Comment by hybridherbst** [comment]
> likely leaving three.js maintainers to deal with the bug reports and possibly implementation of the rest
I'd prefer if something like this was stated instead if implicit, because then there could be a discussion what that rest is in this particular case. In other areas the approach certainly seems to improve piece by piece, instead of covering all known and unknown usecases.

Regarding rotated/scaled parents: my understanding is that this PR ensures the Orbit Controls actually do what the name implies - orbiting around the orbit target. The only thing it changes is that this works, even when the camera doing the orbit is parented to something. It does not change how orbit radius, distance, etc. are applied.

**Comment by looeee** [comment]
@marwie, @hybridherbst and others who feel frustrated by the reasoning behind the rejection of this PR, are you aware of the [three-stdlib](https://github.com/pmndrs/three-stdlib) fork? Or partial fork I guess since it's just the examples folder. 

I believe it was created because many PRs for three.js examples are rejected in a similar manner to this one. You may have more luck submitting new features there. 

For the actual reasoning, which I believe can be stated as: _limit the scope of the examples and keep them as **examples** rather than **production grade plugins**, to reduce the burden on three.js maintainers and keep the focus of this project on improving the core and building support for important features like WebGPU_ - I do think it's a valid reason. You're looking at it from a technical perspective and feeling frustration that a simple fix is rejected (and I have felt this before too). Long time maintainers like @Mugen87 and @donmccurdy are looking at it with the experience of being involved in the day to day running of this repo for many years and understanding that there's only so much that can be done in one unfunded open source project. From this perspective it's important to keep a tight rein on scope creep.

Both are valid perspectives but running this project in a way that maintains its focus has to take priority. If your PR gets rejected with this reasoning, and you still think your code is important enough to share, you should host it elsewhere, like the stdlib repo or in this case see if @yomotsu's camera-controls already supports this. 

As for the responsibility of the maintainers here, I think it's important to acknowledge that this issue does keep coming up and causing frustration for contributors because, as @donmccurdy says, what will or won't be accepted is often a "grey area".  What can we do to communicate this better and make it easier for new contributors to feel welcome here?

**Comment by Mugen87** [comment]
> What can we do to communicate this better and make it easier for new contributors to feel welcome here?

Maybe updating the contribution guide? The note at the end of the documentation is already somewhat related. Maybe it can be extended?

https://github.com/mrdoob/three.js/blob/dev/.github/CONTRIBUTING.md

**Comment by Mugen87** [comment]
> rather than production grade plugins

I wouldn't say that. That implies the modules can't be used in production at all (which is simply not true).

**Comment by CodyJasonBennett** [comment]
> @marwie, @hybridherbst and others who feel frustrated by the reasoning behind the rejection of this PR, are you aware of the [three-stdlib](https://github.com/pmndrs/three-stdlib) fork? Or partial fork I guess since it's just the examples folder.

FWIW, that project first existed because of friction with using examples in modern toolchains (blaming Node CJS/ESM here) and ongoing issues with tree-shaking. I've been looking into backporting fixes here, but aside from tree-shaking, I don't believe it's possible without a build step for examples (and that's already an area I don't want to touch with services like Skypack failing). I did try to suggest similar for three addons for a three-styled bundle, but I realize this is a bit offtopic.

> I believe it was created because many PRs for three.js examples are rejected in a similar manner to this one. You may have more luck submitting new features there.

Notably, three-stdlib had module conversions to Typescript, but it's 1-1 with examples' features (or behind) except exactly OrbitControls with side-effects removed from all controls (see #20575). I believe these frustrations are almost exclusive to controls, in the case of side-effects and general design for OrbitControls to expose internal methods and (seemingly) bug fixes like this one. Part of this, I think, is because of how widely used OrbitControls is, which is why my immediate comparison is GLTFLoader, but I don't know how easy that is for people to contribute to or modify for their own purposes to make a good comparison.

> For the actual reasoning, which I believe can be stated as: limit the scope of the examples and keep them as examples rather than production grade plugins, to reduce the burden on three.js maintainers and keep the focus of this project on improving the core and building support for important features like WebGPU - I do think it's a valid reason. You're looking at it from a technical perspective and feeling frustration that a simple fix is rejected (and I have felt this before too)

I don't like how inconsistent this is because I could go from touching what would otherwise be educational material to something that extends the core and is held to a much higher standard. I hinted at this above as possible misuse, but people use these modules as-is in production and come here if they see a wide enough issue/use-case to fix or support. I think there's a larger organizational issue at hand here or at least something to consider or clarify regarding the intent/organization of the current examples (in regards to scope as per your comment) or general project (with how this is laid out on Github with access to https://github.com/threejs). I wonder how much overhead maintaining these examples in lockstep with the core adds, but I understand the decision to keep distribution unified.

> As for the responsibility of the maintainers here, I think it's important to acknowledge that this issue does keep coming up and causing frustration for contributors because, as @donmccurdy says, what will or won't be accepted is often a "grey area". What can we do to communicate this better and make it easier for new contributors to feel welcome here?

I have no issue accepting this in [three-stdlib](https://github.com/pmndrs/three-stdlib), but I'm hesitant to say that would be of a wider help to this project. People have been pretty vocal here on numerous issues but it's awkward that people feel they have to resort to a community-maintained fork (originally for tooling compat) instead of the official threejs OrbitControls to get enhancements and bug fixes across the line.

People don't feel welcome maybe because they're often not (in the case of OrbitControls) or for reasons that aren't immediately clear to them. It's frustrating to send in a bug fix from a design issue raised multiple times until it was explicitly documented as "wontfix" for it to be shut down as intentional design -- no less, an edge case. Seems there's already a long history here. As a prerequisite to a PR, I might have looked for any prior decisions made in issues or documentation but this specifically is confusing from an outsider's perspective.

> How about updating the contribution guide? The note at the end of the documentation is already somewhat related. Maybe it can be extended?

Would this be good to have on the website? I know the contributing file already follows existing common convention, but wonder if it's worthwhile to have this searchable in the docs.

**Comment by hybridherbst** [comment]
Why are devs bothering with doing PRs?
Because it's better to fix the cause of an issue than to
- patch it up, 
- introduce another dependency,
- introduce duplicated code,
- introduce a fork.
None of the above helps other people, the bug is still present then, and someone else will stumble upon it, as can be seen from the many related PRs and issues here. I hope it's clear that we're trying to contribute. This is not some random product-specific addition.

@Mugen87 would you mind pointing me to the place in the Contributing doc where differences between contributing to core and examples are explained? I didn't find that upon re-reading. Or is that what you'd want to add? 
If so, would that be something like "don't expect fixes to examples being accepted, fixing them would make them less example-y", or "do not make PRs that improve the code in /examples or introduce new usecases", or "please note that PRs will only be accepted if they remove lines"? Sorry if that's tongue-in-cheek but I'd really like to understand this problem, and right now I don't.

I'm still not sure what the disagreement here is, to be honest.
1. The problem fixed by this PR is not a bug?
2. This PR is too complex?
3. This PR would introduce new problems and increase maintenance burden in the future? If so, in which way?

If the only outcome of answering the above is improving the Contribution guide so be it...

**Comment by arodic** [comment]
@marwie @hybridherbst I'm totally with you on this issue. There has been multiple times where I had to make custom OrbitControls to work around this bug. Speaking from experience, I think it is best to maintain your own version - it will give you a lot more freedom and it is better for your mental health.

Contributing to /examples and controls specifically has been one of the biggest challenges in my career. There are so many ways it can be improved but even the most obvious ones often hit into an impenetrable argument of "this is /examples not the /core". That is where many discussions ended in the past just like this one.

**Comment by mqnc** [comment]
Sorry to dig this up, just got screwed by this as well when I wanted to use OrbitControls to adjust the camera in a drone simulator. Current behavior feels absolutely buggy and camera parent should be considered by default.

**Pull request closed without merging**
